crates/astria-conductor/src/sequencer/mod.rs

```diff
@@ -0,0 +1,252 @@
+//! [`Reader`] reads reads blocks from sequencer and forwards them to [`crate::executor::Executor`].
+
+use std::{
+    future::Future,
+    pin::Pin,
+};
+
+use astria_sequencer_types::SequencerBlockData;
+use color_eyre::eyre::{
+    self,
+    bail,
+    WrapErr as _,
+};
+use deadpool::managed::Pool;
+use futures::{
+    future::{
+        self,
+        Ready,
+    },
+    stream::{
+        self,
+        FuturesOrdered,
+    },
+};
+use sequencer_client::{
+    extension_trait::NewBlocksStream,
+    NewBlockStreamError,
+};
+use tokio::{
+    select,
+    sync::oneshot,
+};
+use tracing::{
+    error,
+    info,
+    instrument,
+    warn,
+};
+
+use crate::{
+    client_provider::ClientProvider,
+    executor,
+    executor::ExecutorCommand,
+};
+
+mod sync;
+
+type ResubFut =
+    future::Fuse<Pin<Box<dyn Future<Output = eyre::Result<NewBlocksStream>> + Send + 'static>>>;
+
+pub(crate) struct Reader {
+    /// The channel used to send messages to the executor task.
+    executor_tx: executor::Sender,
+
+    /// The start height from which to start syncing sequencer blocks.
+    start_sync_height: u32,
+
+    /// The object pool providing clients to the sequencer.
+    pool: Pool<ClientProvider>,
+
+    /// The shutdown channel to notify `Reader` to shut down.
+    shutdown: oneshot::Receiver<()>,
+
+    /// The sync-done channel to notify `Conductor` that `Reader` has finished syncing.
+    sync_done: oneshot::Sender<()>,
+}
+
+impl Reader {
+    pub(crate) fn new(
+        start_sync_height: u32,
+        pool: Pool<ClientProvider>,
+        shutdown: oneshot::Receiver<()>,
+        executor_tx: executor::Sender,
+        sync_done: oneshot::Sender<()>,
+    ) -> Self {
+        Self {
+            start_sync_height,
+            executor_tx,
+            pool,
+            shutdown,
+            sync_done,
+        }
+    }
+
+    #[instrument(skip_all)]
+    pub(crate) async fn run_until_stopped(self) -> eyre::Result<()> {
+        use futures::{
+            future::FusedFuture as _,
+            stream::FuturesOrdered,
+            FutureExt as _,
+            StreamExt as _,
+        };
+        let Self {
+            executor_tx,
+            start_sync_height,
+            pool,
+            mut shutdown,
+            sync_done,
+        } = self;
+
+        let mut new_blocks: stream::Fuse<NewBlocksStream> = subscribe_new_blocks(pool.clone())
+            .await
+            .wrap_err("failed to start initial new-blocks subscription")?
+            .fuse();
+
+        let mut pending_blocks = FuturesOrdered::new();
```

Comment:
it appears only `Ready` futures are ever put in here, which is then used in the `select` statement, what's the purpose of this over just calling `forward_pending_block` directly?

Reply:
`pending_blocks` is populated from the new-blocks subscription. However, these are only forwarded after the sync has finished, see the arm in the `select!` macro: `Some(res) = pending_blocks.next(), if sync.is_terminated()`